runtime/devnet/src/config/api/versioning.rs

```diff
@@ -0,0 +1,216 @@
+
+type Version = u8;
+
+/// Versioned runtime calls.
+#[derive(Decode, Debug)]
+pub enum VersionedRuntimeCall {
+	/// Version zero of runtime calls.
+	#[codec(index = 0)]
+	V0(RuntimeCall),
+}
+
+impl From<VersionedRuntimeCall> for RuntimeCall {
+	fn from(value: VersionedRuntimeCall) -> Self {
+		// Allows mapping from some previous runtime call shape to a current valid runtime call
+		match value {
+			VersionedRuntimeCall::V0(call) => call,
+		}
+	}
+}
+
+/// Versioned runtime state reads.
+#[derive(Decode, Debug)]
+pub enum VersionedRuntimeRead {
+	/// Version zero of runtime state reads.
+	#[codec(index = 0)]
+	V0(RuntimeRead),
+}
+
+impl From<VersionedRuntimeRead> for RuntimeRead {
+	fn from(value: VersionedRuntimeRead) -> Self {
+		// Allows mapping from some previous runtime read shape to a current valid runtime read
+		match value {
+			VersionedRuntimeRead::V0(read) => read,
+		}
+	}
+}
+
+/// Versioned runtime state read results.
+#[derive(Debug)]
+pub enum VersionedRuntimeResult {
+	/// Version zero of runtime read results.
+	V0(RuntimeResult),
+}
+
+impl From<(RuntimeResult, Version)> for VersionedRuntimeResult {
+	fn from(value: (RuntimeResult, Version)) -> Self {
+		let (result, version) = value;
+		match version {
+			// Allows mapping from current `RuntimeResult` to a specific/prior version
+			0 => VersionedRuntimeResult::V0(result),
+			// TODO: should never occur due to version processing/validation when request received
+			_ => unimplemented!(),
+		}
+	}
+}
+
+impl From<VersionedRuntimeResult> for Vec<u8> {
+	fn from(result: VersionedRuntimeResult) -> Self {
+		match result {
+			// Simply unwrap and return the encoded result
+			VersionedRuntimeResult::V0(result) => result.encode(),
+		}
+	}
+}
+
+/// Versioned errors.
+#[derive(Debug)]
+pub enum VersionedError {
+	/// Version zero of errors.
+	V0(pop_primitives::v0::Error),
+}
+
+impl From<(DispatchError, Version)> for VersionedError {
+	fn from(value: (DispatchError, Version)) -> Self {
+		let (error, version) = value;
+		match version {
+			// Allows mapping from current `DispatchError` to a specific/prior version of `Error`
+			0 => VersionedError::V0(V0Error::from(error).0),
+			// TODO: should never occur due to version processing/validation when request received
+			_ => unimplemented!(),
+		}
+	}
+}
+
+impl From<VersionedError> for u32 {
+	fn from(value: VersionedError) -> Self {
+		match value {
+			VersionedError::V0(error) => error.into(),
+		}
+	}
+}
+
+struct V0Error(pop_primitives::v0::Error);
+impl From<DispatchError> for V0Error {
+	fn from(error: DispatchError) -> Self {
+		use pop_primitives::v0::*;
+		use sp_runtime::{ArithmeticError::*, TokenError::*, TransactionalError::*};
+		use DispatchError::*;
+		// Mappings exist here to avoid taking a dependency of sp_runtime on pop-primitives
+		Self(match error {
+			Other(_message) => {
+				// Note: lossy conversion: message not used due to returned contract status code size limitation
```

Comment:
Not for this PR, but it would be nice to have a way to write this to the contract buffer for debugging purposes. Maybe we can write to buffer, and then in the contract-side we can read and display when in debug mode only. Needs some more thought.

Reply:
This may actually not be necessary if we return the DispatchError directly for infrastructure-errors (not pop api  use-cases)